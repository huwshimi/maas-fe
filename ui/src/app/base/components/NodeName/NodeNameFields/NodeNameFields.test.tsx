import { mount } from "enzyme";
import { Formik } from "formik";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";

import NodeNameFields from "./NodeNameFields";

import type { RootState } from "app/store/root/types";
import {
  domainState as domainStateFactory,
  machineDetails as machineDetailsFactory,
  machineState as machineStateFactory,
  rootState as rootStateFactory,
} from "testing/factories";

const mockStore = configureStore();

describe("NodeNameFields", () => {
  let state: RootState;
  beforeEach(() => {
    state = rootStateFactory({
      domain: domainStateFactory({
        loaded: true,
      }),
      machine: machineStateFactory({
        loaded: true,
        items: [
          machineDetailsFactory({
            locked: false,
            permissions: ["edit"],
            system_id: "abc123",
          }),
        ],
      }),
    });
  });

  it("displays a spinner when loading domains", () => {
    state.domain.loaded = false;
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Formik
            initialValues={{
              domain: "",
              hostname: "",
            }}
            onSubmit={jest.fn()}
          >
            <NodeNameFields setHostnameError={jest.fn()} />
          </Formik>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Spinner").exists()).toBe(true);
  });

  it("displays the fields", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Formik
            initialValues={{
              domain: "",
              hostname: "",
            }}
            onSubmit={jest.fn()}
          >
            <NodeNameFields setHostnameError={jest.fn()} />
          </Formik>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("NodeNameFields")).toMatchSnapshot();
  });

  it("disables fields when saving", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Formik
            initialValues={{
              domain: "",
              hostname: "",
            }}
            onSubmit={jest.fn()}
          >
            <NodeNameFields saving setHostnameError={jest.fn()} />
          </Formik>
        </MemoryRouter>
      </Provider>
    );
    expect(
      wrapper.find("FormikField").everyWhere((n) => Boolean(n.prop("disabled")))
    ).toBe(true);
  });

  it("updates the hostname errors if they exist", () => {
    const setHostnameError = jest.fn();
    const store = mockStore(state);
    mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <Formik
            initialErrors={{ hostname: "Uh oh!" }}
            initialValues={{
              domain: "",
              hostname: "",
            }}
            onSubmit={jest.fn()}
          >
            <NodeNameFields saving setHostnameError={setHostnameError} />
          </Formik>
        </MemoryRouter>
      </Provider>
    );
    expect(setHostnameError).toHaveBeenCalledWith("Uh oh!");
  });
});
