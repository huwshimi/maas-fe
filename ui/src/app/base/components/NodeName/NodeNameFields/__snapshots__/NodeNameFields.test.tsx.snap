// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeNameFields displays the fields 1`] = `
<NodeNameFields
  setHostnameError={
    [MockFunction] {
      "calls": Array [
        Array [
          null,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="node-name__hostname-wrapper u-no-margin--right"
  >
    <div
      aria-hidden="true"
      className="node-name__hostname-spacer"
    />
    <FormikField
      className="node-name__hostname"
      displayError={false}
      label="Hostname"
      name="hostname"
      takeFocus={true}
      type="text"
      wrapperClassName="u-no-margin--right"
    >
      <Input
        aria-label="Hostname"
        className="node-name__hostname"
        error={null}
        id="mock-redux-js-nanoid-1"
        label="Hostname"
        name="hostname"
        onBlur={[Function]}
        onChange={[Function]}
        takeFocus={true}
        type="text"
        value=""
        wrapperClassName="u-no-margin--right"
      >
        <Field
          className="u-no-margin--right"
          error={null}
          forId="mock-redux-js-nanoid-1"
          helpId="mock-nanoid-2"
          isTickElement={false}
          label="Hostname"
          validationId="mock-nanoid-1"
        >
          <div
            className="p-form__group p-form-validation u-no-margin--right"
          >
            <Label
              forId="mock-redux-js-nanoid-1"
            >
              <label
                className="p-form__label"
                htmlFor="mock-redux-js-nanoid-1"
              >
                Hostname
              </label>
            </Label>
            <div
              className="p-form__control u-clearfix"
            >
              <input
                aria-describedby={null}
                aria-errormessage={null}
                aria-invalid={false}
                aria-label="Hostname"
                className="p-form-validation__input node-name__hostname"
                id="mock-redux-js-nanoid-1"
                label="Hostname"
                name="hostname"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
        </Field>
      </Input>
    </FormikField>
  </div>
  <span
    className="node-name__dot u-nudge-right--small u-nudge-left--small u-no-margin--right"
  >
    .
  </span>
  <DomainSelect
    className="u-no-margin--bottom"
    label="Domain"
    name="domain"
    valueKey="id"
    wrapperClassName="node-name__domain"
  >
    <FormikField
      className="u-no-margin--bottom"
      component={[Function]}
      disabled={false}
      label="Domain"
      name="domain"
      options={
        Array [
          Object {
            "disabled": true,
            "label": "Select domain",
            "value": "",
          },
        ]
      }
      wrapperClassName="node-name__domain"
    >
      <Select
        aria-label="Domain"
        className="u-no-margin--bottom"
        disabled={false}
        error={null}
        id="mock-redux-js-nanoid-2"
        label="Domain"
        name="domain"
        onBlur={[Function]}
        onChange={[Function]}
        options={
          Array [
            Object {
              "disabled": true,
              "label": "Select domain",
              "value": "",
            },
          ]
        }
        value=""
        wrapperClassName="node-name__domain"
      >
        <Field
          className="node-name__domain"
          error={null}
          forId="mock-redux-js-nanoid-2"
          helpId="mock-nanoid-4"
          isSelect={true}
          label="Domain"
          validationId="mock-nanoid-3"
        >
          <div
            className="p-form__group p-form-validation node-name__domain"
          >
            <Label
              forId="mock-redux-js-nanoid-2"
            >
              <label
                className="p-form__label"
                htmlFor="mock-redux-js-nanoid-2"
              >
                Domain
              </label>
            </Label>
            <div
              className="p-form__control u-clearfix"
            >
              <div
                className="p-form-validation__select-wrapper"
              >
                <select
                  aria-describedby={null}
                  aria-errormessage={null}
                  aria-invalid={false}
                  aria-label="Domain"
                  className="p-form-validation__input u-no-margin--bottom"
                  disabled={false}
                  id="mock-redux-js-nanoid-2"
                  name="domain"
                  onBlur={[Function]}
                  onChange={[Function]}
                  value=""
                >
                  <option
                    disabled={true}
                    key="Select domain"
                    value=""
                  >
                    Select domain
                  </option>
                </select>
              </div>
            </div>
          </div>
        </Field>
      </Select>
    </FormikField>
  </DomainSelect>
</NodeNameFields>
`;
